From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 3 Jan 2024 22:57:43 +0100
Subject: [PATCH] Broadcast packets based on player

Adds a method analogous to `PlayerList.broadcast` that allows sending a different packet to each target player.

Adds no usages of this method; those are added in subsequent patches.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index a35638a92479b90afa89cf201fc45b49c9e767f3..cb66ca41f6d527bb9bd50a456bd7a3ad5bb6c77a 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1201,6 +1201,12 @@ public abstract class PlayerList {
     }
 
     public void broadcast(@Nullable net.minecraft.world.entity.player.Player player, double x, double y, double z, double distance, ResourceKey<Level> worldKey, Packet<?> packet) {
+        // Fiddle start - broadcast packets based on player
+        this.broadcastDependentOnPlayer(player, x, y, z, distance, worldKey, targetPlayer -> packet);
+    }
+
+    public void broadcastDependentOnPlayer(@Nullable net.minecraft.world.entity.player.Player player, double x, double y, double z, double distance, ResourceKey<Level> worldKey, Function<ServerPlayer, Packet<?>> packet) {
+        // Fiddle end - broadcast packets based on player
         for (int i = 0; i < this.players.size(); ++i) {
             ServerPlayer entityplayer = (ServerPlayer) this.players.get(i);
 
@@ -1216,7 +1222,7 @@ public abstract class PlayerList {
                 double d6 = z - entityplayer.getZ();
 
                 if (d4 * d4 + d5 * d5 + d6 * d6 < distance * distance) {
-                    entityplayer.connection.send(packet);
+                    entityplayer.connection.send(packet.apply(entityplayer)); // Fiddle end - broadcast packets based on player
                 }
             }
         }
